### ISSUE TITLE
PublicSuffix.fromString should fail gracefully and return null

### ISSUE THREAD

**Issue by benPesso** [Komposten/public_suffix#4]
Often, this method can get called when a URL string is not available yet. It would be great if instead of throwing an exception, it would fail gracefully. Or at least accept an empty string as a valid URL and just "fail" that way.

Without this, we are forced to wrapping `PublicSuffix.fromString` in a `try..catch` or having a valid URL placeholder (which is even worse).

Currently, the below will throw an exception - `The URL is missing the authority component` - if `url` is not only a string, but a valid URL as well. Instead, it should fail gracefully and allow the `Text` widget to accept the empty string as a fallback.
```dart
Text(
  PublicSuffix.fromString(url)?.punyDecoded?.domain ?? '',
  style: TextStyle(
    color: CustomColors.tertiaryText,
    fontSize: 14,
  ),
);
```

**Comment by Komposten** [comment]
This sounds reasonable to me.

However, as it is impossible to extract a suffix from a URL without an authority component, it makes sense in my mind that the default operation should throw (because you are passing invalid input). But, I could add an optional flag or default value that can be used if a graceful failure is preferred.

**Comment by benPesso** [comment]
I agree it should fail without a valid URL, but it should fail and return `null`. Not throw an exception.
You can change the constructor to a factory, though, and it'll still fail when an invalid URL is provided (and fail gracefully when no URL is provided):
```dart
factory PublicSuffix1.fromString(String url) =>
    (url?.isEmpty ?? true) ? null : PublicSuffix1(Uri.parse(url));
```

**Comment by benPesso** [comment]
Hey @Komposten, so should I just fork it and add it on my own branch, or is this something you're planning on implementing?

**Comment by Komposten** [comment]
I've been considering a few different options over the last couple of days but I haven't had time to implement anything. I can push a solution for it during the weekend (most likely tomorrow).

**Comment by Komposten** [comment]
I've pushed a fix to this as part of version 2.0.0. Since I had been working on that version on and off (mostly off) for a while I figured I might as well finish it.

Since this is a major release there are a few other breaking changes. You can check the changelog for all of them, but for the most part you should be good if you just replace your usages of `SuffixRules` with `DefaultSuffixRules`.

**Comment by benPesso** [comment]
Excellent! Love the new changes.
